libs/game/sprite.ts

```diff
@@ -214,18 +215,103 @@ class Sprite implements SpriteLike {
-            this._sayExpires = -1;
-        else
-            this._sayExpires = control.millis() + millis;
+    say(text: string, timeOnScreen?: number, textColor = 15, textBoxColor = 1) {
+        if (!text) {
+            this.updateSay = undefined;
+            return;
+        }
+
+        if (timeOnScreen) {
+            timeOnScreen = timeOnScreen + control.millis();
+        }
+        let pixelsOffset = 0;
+        let holdTextTimer = 1.5;
+        let bubblePadding = 4;
+        let maxTextWidth = 100;
+        let font = image.font8;
+        let startX = 2;
+        let startY = 2;
+        let bubbleWidth = text.length * font.charWidth + bubblePadding;
+        let maxOffset = text.length * font.charWidth - maxTextWidth;
+        let spriteTop = this._hitboxes[0].top;
+
+        for (let i = 0; i < this._hitboxes.length; i++) {
+            if (this._hitboxes[i].top < spriteTop) {
+                spriteTop = this._hitboxes[i].top;
+            }
+        }
+
+        spriteTop = this.y - spriteTop;
+
+        if (bubbleWidth > maxTextWidth + bubblePadding) {
+            bubbleWidth = maxTextWidth + bubblePadding;
+        } else {
+            maxOffset = -1;
+        }
+
+        this.bubbleBoxSprite = sprites.create(image.create(bubbleWidth, font.charHeight + bubblePadding));
+
+        this.bubbleBoxSprite.setFlag(SpriteFlag.Ghost, true);
+
+        this.updateSay = () => {
+            // update box stuff as long as timeOnScreen doesn't exist or it can still be on the screen
+            if (!timeOnScreen || timeOnScreen > control.millis()) {
+                this.bubbleBoxSprite.image.fill(textBoxColor);
+                // the minus 2 is how far the bottom of bubbleBoxSprite to sprites top is
+                this.bubbleBoxSprite.y = this.y - spriteTop - ((font.charHeight + bubblePadding) >> 1) - 2;
+                this.bubbleBoxSprite.x = this.x;
+                // pauses at beginning of text for holdTextTimer length
+                if (holdTextTimer > 0) {
+                    holdTextTimer = holdTextTimer - game.eventContext().deltaTime;
+                    if (holdTextTimer <= 0 && pixelsOffset > 0) {
+                        pixelsOffset = 0;
+                        holdTextTimer = 1.5;
+                    }
+                } else {
+                    pixelsOffset++;
+                    // pause at end of text for holdTextTimer length
+                    if (pixelsOffset >= maxOffset) {
+                        pixelsOffset = maxOffset;
+                        holdTextTimer = 1.5;
+                    }
+                }
+                // if maxOffset is negative it won't scroll
+                if (maxOffset < 0) {
+                    this.bubbleBoxSprite.image.print(text, startX, startY, textColor, font);
+                } else {
+                    this.bubbleBoxSprite.image.print(text, startX - pixelsOffset, startY, textColor, font);
+                }
+
+                // left side padding
```

Comment:
Instead of using for-loops for padding you could just use the `image.fillRect()` API